### DISCUSSION TITLE
How do I get the raw body of a Serverless Function?

### DISCUSSION THREAD

**Discussion by elitan** [nhost/nhost#770] [Tips and Tricks]
The raw body is available at `req.rawBody`:

```js
const handler = (req, res) => {
  const rawBody = req.rawBody

  console.log('rawBody:', rawBody)

  res.send('ok')
}

export default handler
```

To get the types working correctly in TypeScript you can do this:

```js
import { Request, Response } from 'express'

interface NhostRequest extends Request {
  rawBody: string
}

const handler = (req: NhostRequest, res: Response) => {
  const rawBody = req.rawBody

  console.log('rawBody:', rawBody)

  res.send('ok')
}

export default handler
```

And here's an example of how you can use it to set up a webhook for Stripe:

```js
import { Request, Response } from 'express'
import { stripe } from './_utils/stripe'

interface NhostRequest extends Request {
  rawBody: string
}

const endpointSecret = 'whsec_xxx'

const handler = (req: NhostRequest, res: Response) => {
  const sig = req.headers['stripe-signature'] as string

  let event

  try {
    event = stripe.webhooks.constructEvent(req.rawBody, sig, endpointSecret)
  } catch (err: any) {
    console.log(`⚠️  Webhook signature verification failed.`)
    console.log(err)
    return res.status(400).send(`Webhook Error: ${err.message}`)
  }

  if (!event) {
    console.log('no event found')
    console.log(event)
    return res.status(400).send('No event')
  }

  // Handle the event
  switch (event.type) {
    case 'payment_intent.succeeded':
      const paymentIntent = event.data.object
      console.log('PaymentIntent was successful!')
      break
    case 'payment_method.attached':
      const paymentMethod = event.data.object
      console.log('PaymentMethod was attached to a Customer!')
      break
    case 'customer.created':
      const customerCreated = event.data.object
      console.log('customer created!')
      console.log(event.data.object)
      break
    default:
      console.log(`Unhandled event type ${event.type}`)
  }

  res.json({ received: true })
}

export default handler
```

**Comment by wollerman** [comment]
Did something change with `rawBody`? This works great on local! But it fails with a deployed function. I'm doing the same steps here and  `stripe.webhooks.constructEvent(req.rawBody, sig, endpointSecret)` always throws.

Any suggestions?

**Reply by dbarrosop** [reply]
what do you get if you just console log `req.rawBody`?

**Reply by wollerman** [reply]
Here's a `console.log` of both `req.body` vs `req.rawBody`:

regular body
>{"log":"{\n id: 'evt_3NJcRtKE0B3nGMSO1vSmWqHz',\n object: 'event',\n api_version: '2022-11-15',\n created: 1686921143,\n data: {\n object: {\n id: 'pi_3NJcRtKE0B3nGMSO1LmF8nBV',\n object: 'payment_intent',\n amount: 299,\n amount_capturable: 0,\n amount_details: [Object],\n amount_received: 299,\n application: null,\n application_fee_amount: null,\n automatic_payment_methods: [Object],\n canceled_at: null,\n cancellation_reason: null,\n capture_method: 'automatic',\n client_secret: 'redacted',\n confirmation_method: 'automatic',\n created: 1686921141,\n currency: 'usd',\n customer: 'redacted',\n description: 'redacted',\n invoice: null,\n last_payment_error: null,\n latest_charge: 'ch_3NJcRtKE0B3nGMSO1hRQkuEM',\n livemode: false,\n metadata: [Object],\n next_action: null,\n on_behalf_of: null,\n payment_method: 'pm_1NJbSAKE0B3nGMSOCNC0m6Xq',\n payment_method_options: [Object],\n payment_method_types: [Array],\n processing: null,\n receipt_email: 'redacted@gmail.com',\n review: null,\n setup_future_usage: null,\n shipping: null,\n source: null,\n statement_descriptor: null,\n statement_descriptor_suffix: null,\n status: 'succeeded',\n transfer_data: null,\n transfer_group: null\n }\n },\n livemode: false,\n pending_webhooks: 1,\n request: {\n id: 'req_j4SoXVHCLk8aF9',\n idempotency_key: 'redacted'\n },\n type: 'payment_intent.succeeded'\n}","path":"/payment/stripe_hook","requestId":"51934a7d-ede9-4537-889d-231609aa6efd","level":"INFO"}

raw body
>{"log":"{\n \"id\": \"evt_3NJcRtKE0B3nGMSO1vSmWqHz\",\n \"object\": \"event\",\n \"api_version\": \"2022-11-15\",\n \"created\": 1686921143,\n \"data\": {\n \"object\": {\n \"id\": \"pi_3NJcRtKE0B3nGMSO1LmF8nBV\",\n \"object\": \"payment_intent\",\n \"amount\": 299,\n \"amount_capturable\": 0,\n \"amount_details\": {\n \"tip\": {\n }\n },\n \"amount_received\": 299,\n \"application\": null,\n \"application_fee_amount\": null,\n \"automatic_payment_methods\": {\n \"allow_redirects\": \"always\",\n \"enabled\": true\n },\n \"canceled_at\": null,\n \"cancellation_reason\": null,\n \"capture_method\": \"automatic\",\n \"client_secret\": \"redacted\",\n \"confirmation_method\": \"automatic\",\n \"created\": 1686921141,\n \"currency\": \"usd\",\n \"customer\": \"redacted\",\n \"description\": \"redacted\",\n \"invoice\": null,\n \"last_payment_error\": null,\n \"latest_charge\": \"ch_3NJcRtKE0B3nGMSO1hRQkuEM\",\n \"livemode\": false,\n \"metadata\": {\n \"tokens\": \"10\",\n \"userId\": \"0291f8fc-9698-4792-89fb-cca9c04e1e84\"\n },\n \"next_action\": null,\n \"on_behalf_of\": null,\n \"payment_method\": \"pm_1NJbSAKE0B3nGMSOCNC0m6Xq\",\n \"payment_method_options\": {\n \"card\": {\n \"installments\": null,\n \"mandate_options\": null,\n \"network\": null,\n \"request_three_d_secure\": \"automatic\"\n },\n \"cashapp\": {\n },\n \"link\": {\n \"persistent_token\": null\n }\n },\n \"payment_method_types\": [\n \"card\",\n \"link\",\n \"cashapp\"\n ],\n \"processing\": null,\n \"receipt_email\": \"redacted@gmail.com\",\n \"review\": null,\n \"setup_future_usage\": null,\n \"shipping\": null,\n \"source\": null,\n \"statement_descriptor\": null,\n \"statement_descriptor_suffix\": null,\n \"status\": \"succeeded\",\n \"transfer_data\": null,\n \"transfer_group\": null\n }\n },\n \"livemode\": false,\n \"pending_webhooks\": 1,\n \"request\": {\n \"id\": \"req_j4SoXVHCLk8aF9\",\n \"idempotency_key\": \"redacted\"\n },\n \"type\": \"payment_intent.succeeded\"\n}","path":"/payment/stripe_hook","requestId":"51934a7d-ede9-4537-889d-231609aa6efd","level":"INFO"}
